app/upgrades.go

```diff
@@ -205,6 +208,17 @@ var upgrades = map[string]appUpgrade{
 		Added: []string{crisistypes.ModuleName},
 		Handler: func(ctx sdk.Context, app *App, vm module.VersionMap) (module.VersionMap, error) {
 			var err error
+
+			err = migrateParams(ctx, app)
+			if err != nil {
+				return nil, err
+			}
+
+			err = upgradeIBC(ctx, app)
+			if err != nil {
+				return nil, err
+			}
+
```

Comment:
Usually the custom migration stuff needs to happen after `runModuleMigrations`. Otherwise, you're using code/functions designed for the new state against data formatted using the old state. Are we sure these should be done first?

Reply:
This is the order that was recommended by IBC. I also noticed that our chain runs module migrations first, but other chains and the sdk documentation recommend doing it last. I'm digging into the documentation for this now, but I don't think our chain does anything exceptional that should differentiate it.

Reply:
Gotcha. I think that, in general, you want stuff to happen after the module migrations, but there are definitely cases where you want things done before them.

What can happen is a discrepancy between how the data is stored and how the data is expected. We had an upgrade a while back where this was an issue, but I can't remember too many details; I believe it was bank module related, though.

Basically, we needed to do some lookup/processing of stuff in a module that had an upgrade. The migration was moving a bunch of module's data in order to use different key and/or value formats. E.g. all the state keys that started with `0x01` were being moved to `0x02`. All the code (except for inside the migration) expected the data to be at `0x02`. So if we did our lookup/processing first, it just looked like there wasn't any data. Worse problems can happen if the key or value formats change without being moved; the new code tries to read the entry using the new format, but can't, resulting in anything from a panic to a silent failure.

A more concrete example is `removeInactiveValidatorDelegations`. If there's a module upgrade in the staking keeper that changes how that stuff is recorded, then we need to do that **after** the module migrations.